compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1835,6 +1836,56 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    pub(super) fn collect_unused_for_coerce_return_ty(
+        &self,
+        errors: &[traits::FulfillmentError<'tcx>],
+    ) {
+        for fillment_error in &*errors {
+            let obligation = &fillment_error.obligation;
+            let body_id = obligation.cause.body_id;
+            let span = obligation.cause.span;
+
+            let Some(body_id) = self.tcx.hir().maybe_body_owned_by(
+                self.tcx.typeck_root_def_id(body_id.to_def_id()).expect_local(),
+            ) else {
+                continue;
+            };
+
+            let Some(mut diag) =
+                self.tcx.sess.diagnostic().steal_diagnostic(span, StashKey::MaybeForgetReturn)
+            else {
+                continue;
+            };
+
+            let body = self.tcx.hir().body(body_id);
+            let root_obligation = &fillment_error.root_obligation;
+            if let Some((_, fn_decl, _)) = self.get_fn_decl(body.value.hir_id)
+                && let hir::FnRetTy::Return(_ty) = fn_decl.output
+                && let Some(fn_sig) = self.body_fn_sig()
+                && let ExprBindingObligation(_, _, hir_id, ..) = root_obligation.cause.code() {
+                    for (_, node) in self.tcx.hir().parent_iter(*hir_id) {
+                        if let hir::Node::Stmt(stmt) = node &&
+                            let hir::StmtKind::Semi(ref expr) = stmt.kind &&
+                                expr.span.contains(span) {
+                                let expr_ty = self.typeck_results.borrow().expr_ty(expr);
+                                let return_ty = fn_sig.output();
+                                if !matches!(expr.kind, hir::ExprKind::Ret(..)) &&
```

Comment:
If we check that the direct parent of the `hir_id` is a semi then this shouldn't be necessary anymore?